File: app/src/main/java/com/startogamu/zikobot/player/alarm/AlarmBottomFragmentVM.java
package com.startogamu.zikobot.player.alarm;

import android.content.Context;
import android.media.AudioManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.widget.SeekBar;

import com.joxad.easydatabinding.bottomsheet.DialogBottomSheetBaseVM;
import com.joxad.zikobot.data.db.AlarmManager;
import com.joxad.zikobot.data.db.model.ZikoAlarm;
import com.startogamu.zikobot.Constants;
import com.startogamu.zikobot.databinding.AlarmFragmentBinding;

/**
 * Created by Jocelyn on 22/08/2017.
 */

public class AlarmBottomFragmentVM extends DialogBottomSheetBaseVM<AlarmBottomFragment, AlarmFragmentBinding> {

    public AlarmVM alarmVM;
    private AudioManager am;

    /***
     * @param fragment
     * @param binding
     */
    public AlarmBottomFragmentVM(AlarmBottomFragment fragment, AlarmFragmentBinding binding) {
        super(fragment, binding);
    }

    @Override
    public void onCreate() {
        ZikoAlarm alarm = AlarmManager.INSTANCE.getAlarmByPlaylistId(fragment.getArguments().getLong(Constants.Extra.INSTANCE.getPLAYLIST_ID()));
        alarmVM = new AlarmVM(fragment.getContext(), alarm);
        am = (AudioManager) fragment.getActivity().getSystemService(Context.AUDIO_SERVICE);
        init();
    }


    private void init() {
        int max = am.getStreamMaxVolume(AudioManager.STREAM_MUSIC);
        binding.seekBarVolume.setMax(am.getStreamMaxVolume(AudioManager.STREAM_MUSIC));
        binding.seekBarVolume.setProgress(alarmVM.getVolume() == -1 ? ((int) (max * 0.5f)) : alarmVM.getVolume());
        binding.swRandom.setChecked(alarmVM.getRandomTrack() == 1);
        binding.timePicker.setIs24HourView(true);
        initHour();

        binding.swRandom.setOnCheckedChangeListener((compoundButton, b) -> alarmVM.updateRandom(b));

        binding.seekBarVolume.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                if (b)
                    alarmVM.updateVolume(i);
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });
    }

    private void initHour() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            binding.timePicker.setMinute(alarmVM.getMinute());
            binding.timePicker.setHour(alarmVM.getHour());
        } else {
            binding.timePicker.setCurrentHour(alarmVM.getHour());
            binding.timePicker.setCurrentMinute(alarmVM.getMinute());
        }
    }

    @Deprecated
    @Override
    public void onCreate(@Nullable Bundle savedInstance) {

    }
}

File: app/src/main/java/com/startogamu/zikobot/player/alarm/AlarmBottomFragmentVM.kt
package com.startogamu.zikobot.player.alarm

import android.content.Context
import android.media.AudioManager
import android.os.Bundle
import android.view.View
import android.widget.SeekBar
import com.joxad.easydatabinding.bottomsheet.DialogBottomSheetBaseVM
import com.philliphsu.bottomsheetpickers.time.grid.GridTimePickerDialog
import com.philliphsu.bottomsheetpickers.time.numberpad.NumberPadTimePickerDialog
import com.startogamu.zikobot.Constants
import com.startogamu.zikobot.databinding.AlarmFragmentBinding

/**
 * Created by Jocelyn on 22/08/2017.
 */

class AlarmBottomFragmentVM(fragment: AlarmBottomFragment, binding: AlarmFragmentBinding)
    : DialogBottomSheetBaseVM<AlarmBottomFragment, AlarmFragmentBinding>(fragment, binding) {

    lateinit var alarmVM: AlarmVM
    private var am: AudioManager? = null

    /***
     *
     */
    override fun onCreate() {
        val alarm = AlarmManager.INSTANCE.getAlarmByPlaylistId(fragment.arguments.getLong(Constants.Extra.PLAYLIST_ID))
        alarmVM = AlarmVM(fragment.context, alarm)
        am = fragment.activity.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        init()
    }


    /***
     *
     */
    fun save(@SuppressWarnings("unused") v: View) {
        alarmVM.save()
        fragment.dismiss()
    }

    /****
     *
     */
    private fun init() {
        val max = am!!.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        binding.seekBarVolume.max = am!!.getStreamMaxVolume(AudioManager.STREAM_MUSIC)
        binding.seekBarVolume.progress = if (alarmVM.volume == -1) (max * 0.5f).toInt() else alarmVM.volume
        binding.swRandom.isChecked = alarmVM.randomTrack == 1

        binding.swRandom.setOnCheckedChangeListener { _, b -> alarmVM.updateRandom(b) }

        binding.seekBarVolume.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {
                if (b)
                    alarmVM.updateVolume(i)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })
    }

    public fun editTime(v: View) {
        /* if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
             binding.timePicker.minute = alarmVM.minute
             binding.timePicker.hour = alarmVM.hour
         } else {
             binding.timePicker.currentHour = alarmVM.hour
             binding.timePicker.currentMinute = alarmVM.minute
         }
         binding.timePicker.setOnTimeChangedListener { _, i, i1 ->
             alarmVM.updateTimeSelected(currentHour = i, currentMin = i1)
         }*/
        // Configured according to the system preference for 24-hour time.
        val pad = NumberPadTimePickerDialog.Builder({ _, i, i1 ->
            alarmVM.updateTimeSelected(currentHour = i, currentMin = i1)

        },true).setThemeDark(true).build()
        pad.show(fragment.activity.supportFragmentManager, pad.tag)

    }

    @Deprecated("")
    override fun onCreate(savedInstance: Bundle?) {

    }
}
